url20b0f880a5_jgraph_legacy_jgraph5_tgz-pJ8-org_jgraph_example_HelloWorldJ8/src/org/jgraph/graph/DefaultEdge.java

```diff
@@ -0,0 +1,247 @@
+	 * and sets default values for points and the label position.
+	 * 
+	 * @param userObject
+	 *            reference to the user object
+	 */
+	public DefaultEdge(Object userObject, AttributeMap storageMap) {
+		super(userObject, storageMap);
+	}
+
+	/**
+	 * Returns the source of the edge.
+	 */
+	public Object getSource() {
+		return source;
+	}
+
+	/**
+	 * Returns the target of the edge.
+	 */
+	public Object getTarget() {
+		return target;
+	}
+
+	/**
+	 * Sets the source of the edge.
+	 */
+	public void setSource(Object port) {
+		source = port;
+	}
+
+	/**
+	 * Returns the target of <code>edge</code>.
+	 */
+	public void setTarget(Object port) {
+		target = port;
+	}
+
+	/**
+	 * Create a clone of the cell. The cloning of the user object is deferred to
+	 * the cloneUserObject() method. The source and target references are set to
+	 * null.
+	 * 
+	 * @return Object a clone of this object.
+	 */
+	public Object clone() {
+		DefaultEdge edge = (DefaultEdge) super.clone();
+		edge.source = null;
+		edge.target = null;
+		return edge;
+	}
+
+	//
+	// Default Routing
+	// 
+
+	public static class LoopRouting implements Edge.Routing {
+
+		public List route(GraphLayoutCache cache, EdgeView edge) {
+			if (edge.isLoop()) {
+				return routeLoop(cache, edge);
+			}
+			return routeEdge(cache, edge);
+		}
+
+		protected List routeLoop(GraphLayoutCache cache, EdgeView edge) {
+			List newPoints = new ArrayList();
+			newPoints.add(edge.getSource());
+			CellView sourceParent = (edge.getSource() != null) ? edge
+					.getSource().getParentView() : edge.getSourceParentView();
+			if (sourceParent != null) {
+				Point2D from = AbstractCellView.getCenterPoint(sourceParent);
+				Rectangle2D rect = sourceParent.getBounds();
+				double width = rect.getWidth();
+				double height2 = rect.getHeight() / 2;
+				double loopWidth = Math.min(20, Math.max(10, width / 8));
+				double loopHeight = Math.min(30, Math.max(12, Math.max(
+						loopWidth + 4, height2 / 2)));
+				newPoints.add(edge.getAttributes().createPoint(
+						from.getX() - loopWidth,
+						from.getY() - height2 - loopHeight * 1.2));
+				newPoints.add(edge.getAttributes().createPoint(from.getX(),
+						from.getY() - height2 - 1.5 * loopHeight));
+				newPoints.add(edge.getAttributes().createPoint(
+						from.getX() + loopWidth,
+						from.getY() - height2 - loopHeight * 1.2));
+				newPoints.add(edge.getTarget());
+				return newPoints;
+			}
+			return null;
+		}
+
+		protected List routeEdge(GraphLayoutCache cache, EdgeView edge) {
+			return null;
+		}
+
+		public int getPreferredLineStyle(EdgeView edge) {
+			if (edge.isLoop()) {
+				return getLoopStyle();
+			}
+			return getEdgeStyle();
+		}
+
+		protected int getLoopStyle() {
+			return GraphConstants.STYLE_BEZIER;
+		}
+
+		protected int getEdgeStyle() {
+			return NO_PREFERENCE;
+		}
+
+	}
+
+	public static class DefaultRouting extends LoopRouting {
+
+		protected List routeEdge(GraphLayoutCache cache, EdgeView edge) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 22. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.